django/api/tasks.py

```diff
@@ -362,3 +366,36 @@ def check_rebates_redeemed_since(iso_ts=None, schedule_func_name=None):
         status=GoElectricRebateApplication.Status.REDEEMED,
         modified=timezone.now(),
     )
+
+
+# Auto-Cancel "household" applications in "initiated" status for more than 28 days
+def cancel_household_applications_initiated_status():
+    # get all applications in "initiated" status
+    applications = GoElectricRebateApplication.objects.filter(
+        status=GoElectricRebateApplication.Status.HOUSEHOLD_INITIATED
+    )
+
+    # get all applications that are "household" and "initiated" for more than 28 days
+    household_applications_initiated_28_days = applications.filter(
+        modified__lte=timezone.now() - timedelta(days=28)
+    )
+
+    # send email to all household applications that are "initiated" for more than 28 days
+    for application in household_applications_initiated_28_days:
+        print("Sending cancel email to {}".format(application.email))
+        send_cancel(application.email, application.id)
+        print("cancelled " + application.email)
+
+    # cancel all household applications that are "initiated" for more than 28 days
+    household_applications_initiated_28_days.update(
+        status=GoElectricRebateApplication.Status.CANCELLED,
+        modified=timezone.now(),
+    )
```

Comment:
for sending the cancel emails, we emit signals that create async_tasks so that if an email fails, it will retry later. I think what we want to do here is, after updating, manually emit the post_save signal, since the update() function doesn't send that signal automatically (please see update_application_statuses() in rebate.py for example of what can be done). Also, I don't think you need to update the "modified" field; that should be a self-managed field.

Reply:
post_save.send() should occur after the update, and you also need to pass in update_fields=["status"]

Reply:
@tim738745 I have observed that the filtered applications variable "household_applications_initiated_28_days" returns empty once the status id's are updated. Hence, the signal was added earlier. Modified now, but need to test once deployed.

Reply:
@vibhiquartech I guess that's because the actual evaluation of the queryset does not occur until you iterate, and at that point, the statuses have been updated. What you could try is forcing evaluation using list() (or some other method; see https://docs.djangoproject.com/en/4.0/ref/models/querysets/#when-querysets-are-evaluated) before you update, and iterate through those instead. So like: x = list(household_applications_initiated_28_days) and then update, and then iterate through x.